### PR TITLE
add solutions for assignemnt 3

### PR DESCRIPTION
**Pull request by katermar**

### CODE REVIEW

File: assignment3/Dictionary.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package com.katermar.assignment3;
+
+import java.util.HashSet;
+
+/**
+ * Created by katermar on 6/8/2017.
+ */
+public class Dictionary {
```
Comment (brzozan) [inline assignment3/Dictionary.java]: No need to submit Dictionary class, but feel free to leave it here.

And since it's here, I will add some comments :)

File: assignment3/WordFinder.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * Created by katermar on 6/8/2017.
+ */
+
+/*
+class WordFinder to find words in a char grid
+ */
+public class WordFinder {
+    private Dictionary dictionary;
```
Comment (brzozan) [inline assignment3/WordFinder.java]: Could be final, as `setDictionary` is not really needed. It is better to initialize classes once in constructor and keep them immutable afterwards. It reduces code complexity and lets you avoid serious bugs, especially if the class will ever be used in multi-threaded environment.

If you need to find words from a different dictionary in the future, it is cheap to just create new `WordFinder` instance initialized with new Dictionary.

File: assignment3/WordFinder.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    public WordFinder(Dictionary dictionary) {
+        this.dictionary = dictionary;
+    }
+
+    public Dictionary getDictionary() {
```
Comment (brzozan) [inline assignment3/WordFinder.java]: Not needed. It's good to keep interface of the class as small as possible and remove all not needed code. Less things to worry about when maintaining code in the future.

File: assignment3/WordFinder.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        return dictionary;
+    }
+
+    public void setDictionary(Dictionary dictionary) {
```
Comment (brzozan) [inline assignment3/WordFinder.java]: Not needed.

File: assignment3/WordFinder.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            for (int col = j - 1; col <= j + 1 && col < visited[0].length; col++) {
+                if (row >= 0 && col >= 0 && !visited[row][col] && dictionary.isPrefix(sample)) {
+                    findWordsHelper(result, grid, visited, row, col, sample);
+                }
+            }
+        }
+
+        // delete character from string and mark visited
+        // of current cell as false
+        sample = sample.substring(0, sample.length() - 1);
```
Comment (brzozan) [inline assignment3/WordFinder.java]: Not needed. Strings in Java are immutable, so you have not modified the value `sample` has in the method that called (recursively or not) `findWordsHelper`.

File: assignment3/WordFinder.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    /*A recursive function that finds all words in a char grid
+      using dictionary, acts like DFS
+      --------------------------------------------------------
+      @param grid       -- char grid
+      --------------------------------------------------------
+      @return result set of words
+     */
+    public Set<String> findWords(char grid[][]) {
+        //initialize array, which marks all cells non-visited by default
+        boolean visited[][] = new boolean[grid.length][grid[0].length];
```
Comment (brzozan) [inline assignment3/WordFinder.java]: What if someone calls findWords with empty grid?

File: assignment3/WordFinder.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        HashSet<String> result = new HashSet<>();
+        String sample = "";
```
Comment (brzozan) [inline assignment3/WordFinder.java]: Not really needed. You could just pass "" as parameter to findWordsHelper.

File: assignment3/Dictionary.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package com.katermar.assignment3;
+
+import java.util.HashSet;
+
+/**
+ * Created by katermar on 6/8/2017.
+ */
+public class Dictionary {
+    private HashSet<String> dictionary = new HashSet<>();
+    private HashSet<String> prefixes = new HashSet<>();
```
Comment (brzozan) [inline assignment3/Dictionary.java]: Both could be final and initialized in constructor rather than via setters.

File: assignment3/Dictionary.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    public boolean isPrefix(String prefix){
+      return prefixes.contains(prefix);
+    }
+
+    public boolean isWord(String word) {
+        return dictionary.contains(word);
+    }
+
+    public HashSet<String> getDictionary() {
```
Comment (brzozan) [inline assignment3/Dictionary.java]: Not really needed...

File: assignment3/Dictionary.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        return dictionary;
+    }
+
+    public void setDictionary(HashSet<String> dictionary) {
+        this.dictionary = dictionary;
+    }
+
+    public HashSet<String> getPrefixes() {
```
Comment (brzozan) [inline assignment3/Dictionary.java]: Not really needed...

File: assignment3/Dictionary.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    public boolean isWord(String word) {
+        return dictionary.contains(word);
+    }
+
+    public HashSet<String> getDictionary() {
+        return dictionary;
+    }
+
+    public void setDictionary(HashSet<String> dictionary) {
```
Comment (brzozan) [inline assignment3/Dictionary.java]: As mentioned above, would be better to have initialization in constructor.

File: assignment3/Dictionary.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        this.dictionary = dictionary;
+    }
+
+    public HashSet<String> getPrefixes() {
+        return prefixes;
+    }
+
+    public void setPrefixes(HashSet<String> prefixes) {
```
Comment (brzozan) [inline assignment3/Dictionary.java]: Again, initialization would be better in constructor.

Additionally prefixes could be trivially created automatically from the dictionary, saving you manual work.

File: assignment3/WordFinderTest.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        };
+        Dictionary dictionary = new Dictionary();
+        String dictArray[] = {"CAR", "CARD", "CART", "CAT"};
+        String prefixArray[] = {"C", "CA", "CAR", "CARD", "CART", "CAT"};
+        String expectedResultArray[] = {"CAR", "CARD", "CAT"};
+        dictionary.setDictionary(new HashSet<String>(Arrays.asList(dictArray)));
+        dictionary.setPrefixes(new HashSet<String>(Arrays.asList(prefixArray)));
+
+        testWordFinder(new HashSet<String>(Arrays.asList(expectedResultArray)), dictionary, grid);
+        testWordFinder(new HashSet<String>(), dictionary, grid);
```
Comment (brzozan) [inline assignment3/WordFinderTest.java]: Thank you for testing edge case! Could you also add a test for empty grid?

File: assignment3/Dictionary.java [hunk 14]
```diff
@@ -1,35 +1,50 @@
...
-        return dictionary.contains(word);
+    public Dictionary(String[] dictionary, String[] prefixes) {
+        this.dictionary = new HashSet<>(Arrays.asList(dictionary));
+        this.prefixes = new HashSet<>(Arrays.asList(prefixes));
     }
 
-    public HashSet<String> getDictionary() {
-        return dictionary;
+    public Dictionary(String[] dictionary) {
+        this.dictionary = new HashSet<>(Arrays.asList(dictionary));
```
Comment (brzozan) [inline assignment3/Dictionary.java]: There is some code duplication between this constructor and the one accepting HashSet. You could just use constructor chaining here:
```
public Dictionary(String[] dictionary) {
  super(new HashSet<>(Arrays.asList(dictionary));
}
```

File: assignment3/Dictionary.java [hunk 15]
```diff
@@ -1,35 +1,50 @@
...
+        this.prefixes = new HashSet<>();
+        for (String word : dictionary) {
+            for (int i = 0; i < word.length(); i++) {
+                prefixes.add(word.substring(0, i));
+            }
+        }
     }
 
-    public void setDictionary(HashSet<String> dictionary) {
+    public Dictionary(HashSet<String> dictionary) {
```
Comment (brzozan) [inline assignment3/Dictionary.java]: Just FYI comment. Feel free to not the code.

It's better to use interface (`Set`) as parameter type rather than concrete type (`HashSet`). It gives much more flexibility for the users of the class - they could prefer using `TreeSet` or something similar.
To get the performance guarantees of a `HashSet` in your code (and to protect from a malicious user that might change contents of the dictionary set without your knowledge) you would just copy it in constructor.

File: assignment3/WordFinder.java (hunk 6 above)
Comment (brzozan) [inline assignment3/WordFinder.java]: Please add protection against empty grid.

File: assignment3/WordFinder.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+      using dictionary, acts like DFS
+      --------------------------------------------------------
+      @param grid       -- char grid
+      --------------------------------------------------------
+      @return result set of words
+     */
+    public Set<String> findWords(char grid[][]) {
+        //initialize array, which marks all cells non-visited by default
+        boolean visited[][] = new boolean[grid.length][grid[0].length];
+        char checkEmpty[][] = new char[grid.length][grid[0].length];
```
Comment (brzozan) [inline assignment3/WordFinder.java]: Clever, but consider what happens if this method is called with a very large grid. Note, that this method is called recursively, potentially many, many times!

Checking for any element being true using for loops will be much more efficient and will not result in much more code.

File: assignment3/WordFinderTest.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+                {'T', 'C', 'D'}
+        };
+        String dictArray[] = {"CAR", "CARD", "CART", "CAT"};
+//        String prefixArray[] = {"C", "CA", "CAR", "CARD", "CART", "CAT"};
+        Dictionary dictionary = new Dictionary(dictArray);
+        String expectedResultArray[] = {"CAR", "CARD", "CAT"};
+
+        testWordFinder(new HashSet<String>(Arrays.asList(expectedResultArray)), dictionary, grid);
+        testWordFinder(new HashSet<String>(), dictionary, grid);
+        char emptyGrid[][] = {{}, {}};
```
Comment (brzozan) [inline assignment3/WordFinderTest.java]: Can you also add test for even more empty grid, i.e. `{}`?

File: assignment3/WordFinderTest.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class WordFinderTest {
+
+    public static void main(String args[]) {
+        char grid[][] = {
+                {'A', 'A', 'R'},
+                {'T', 'C', 'D'}
+        };
+        String dictArray[] = {"CAR", "CARD", "CART", "CAT"};
+//        String prefixArray[] = {"C", "CA", "CAR", "CARD", "CART", "CAT"};
```
Comment (brzozan) [inline assignment3/WordFinderTest.java]: Usually you do not want to leave commented out code, especially without a `// TODO` comment explaining what it is for. Such code tends to get neglected during maintenance and rots.

File: assignment3/WordFinderTest.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+        Dictionary dictionary = new Dictionary(dictArray);
+        String expectedResultArray[] = {"CAR", "CARD", "CAT"};
+
+        testWordFinder(new HashSet<String>(Arrays.asList(expectedResultArray)), dictionary, grid);
+        testWordFinder(new HashSet<String>(), dictionary, grid);
+        char emptyGrid[][] = {{}, {}};
+        testWordFinder(new HashSet<String>(Arrays.asList(expectedResultArray)), dictionary, emptyGrid);
+    }
+
+    public static void testWordFinder(HashSet<String> expectedResult, Dictionary dictionary, char grid[][]) {
```
Comment (brzozan) [inline assignment3/WordFinderTest.java]: If you define your own testing mechanism it is useful to have tests tagged somehow and print this tag as part of the success/failure message, so that it is trivial to find out which test case failed.
It might not be a problem for three test cases, but if you have hundreds of them and you face a stream of multiple pass/fail messages that you then have to manually count to map them to test cases that need to be fixed.

Could you please add something like test name to the method and print it together with results?

**Comment by brzozan** [comment]
Sorry for the delay. GitHub somehow decided to not let me know that you made changes after my previous review.

LGTM. Thank you for following and for your patience.
